libs/vkd3d/command.c

```diff
@@ -1681,6 +1681,284 @@ static void d3d12_command_list_invalidate_current_pipeline(struct d3d12_command_
+            return VK_IMAGE_ASPECT_COLOR_BIT;
+        case VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL:
+            return VK_IMAGE_ASPECT_DEPTH_BIT | VK_IMAGE_ASPECT_STENCIL_BIT;
+        case VK_IMAGE_LAYOUT_DEPTH_READ_ONLY_STENCIL_ATTACHMENT_OPTIMAL:
+            return VK_IMAGE_ASPECT_STENCIL_BIT;
+        case VK_IMAGE_LAYOUT_DEPTH_ATTACHMENT_STENCIL_READ_ONLY_OPTIMAL:
+            return VK_IMAGE_ASPECT_DEPTH_BIT;
+        case VK_IMAGE_LAYOUT_UNDEFINED:
+        case VK_IMAGE_LAYOUT_DEPTH_STENCIL_READ_ONLY_OPTIMAL:
+            return 0;
+        default:
+            ERR("Unhandled image layout %u.\n", vk_layout);
+            return 0;
+    }
+}
+
+static int d3d12_command_list_find_attachment(struct d3d12_command_list *list, const struct d3d12_resource *resource,
+        const struct vkd3d_view *view, VkImageAspectFlags clear_aspects)
+{
+    unsigned int i;
+
+    if (!list->current_render_pass)
+        return -1;
+
+    if (list->dsv.resource == resource)
+    {
+        const struct vkd3d_view *dsv = list->dsv.view;
+
+        if ((vk_writable_aspects_from_image_layout(list->dsv_layout) & clear_aspects) != clear_aspects)
+            return -1;
+
+        if (dsv->info.texture.miplevel_idx == view->info.texture.miplevel_idx &&
+                dsv->info.texture.layer_idx == view->info.texture.layer_idx &&
+                dsv->info.texture.layer_count == view->info.texture.layer_count)
+            return D3D12_SIMULTANEOUS_RENDER_TARGET_COUNT;
+    }
+    else
+    {
+        for (i = 0; i < D3D12_SIMULTANEOUS_RENDER_TARGET_COUNT; i++)
+        {
+            const struct vkd3d_view *rtv = list->rtvs[i].view;
+
+            if (list->rtvs[i].resource != resource)
+                continue;
+
+            if (rtv->info.texture.miplevel_idx == view->info.texture.miplevel_idx &&
+                    rtv->info.texture.layer_idx == view->info.texture.layer_idx &&
+                    rtv->info.texture.layer_count == view->info.texture.layer_count)
+                return i;
+        }
+    }
+
+    return -1;
+}
+
+static void d3d12_command_list_clear_attachment_inline(struct d3d12_command_list *list, struct d3d12_resource *resource,
+        struct vkd3d_view *view, unsigned int attachment_idx, VkImageAspectFlags clear_aspects,
+        const VkClearValue *clear_value, UINT rect_count, const D3D12_RECT *rects)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &list->device->vk_procs;
+    VkClearAttachment vk_clear_attachment;
+    VkClearRect vk_clear_rect;
+    D3D12_RECT full_rect;
+    unsigned int i;
+
+    if (!rect_count)
+    {
+        full_rect = d3d12_get_view_rect(resource, view);
+        rect_count = 1;
+        rects = &full_rect;
+    }
+
+    /* We expect more than one clear rect to be very uncommon
+     * in practice, so make no effort to batch calls for now. */
+    vk_clear_attachment.aspectMask = clear_aspects;
+    vk_clear_attachment.colorAttachment = attachment_idx;
```

Comment:
Does this work with depth-stencil?

Reply:
Yes, the color attachment index must be ignored if `clear_aspects` contains depth or stencil.

Could set this to 0, but even 0 isn't necessarily valid either (think depth-only passes).

Reply:
Could just add a comment to make it clear.

Reply:
Done.